membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -38,12 +38,37 @@ class AccountController extends LazyLogging {
   lazy val mmaAction = NoCacheAction andThen mmaCorsFilter andThen BackendFromCookieAction
   lazy val mmaCardAction = NoCacheAction andThen corsCardFilter andThen BackendFromCookieAction
 
+
+  def cancelSubscription [P <: SubscriptionPlan.AnyPlan : SubPlanReads] = mmaCardAction.async { implicit request =>
+    val tp = request.touchpoint
+    val cancelForm = Form { single("reason" -> nonEmptyText) }
+
+    val maybeUserId = authenticationService.userId
+    logger.info(s"Attempting to cancel contribution for $maybeUserId")
+
+    (for {
+      user <- EitherT(Future.successful( maybeUserId \/> "no identity cookie for user"))
+      sfUser <- EitherT(tp.contactRepo.get(user).map(_.flatMap(_ \/> s"no SF user $user")))
+      zuoraSubscription <- EitherT(tp.subService.current[P](sfUser).map(_.headOption).map (_ \/> s"no current subscriptions for the sfUser $sfUser"))
+      _ <- EitherT(tp.zuoraRestService.disableAutoPay(zuoraSubscription.accountId)).leftMap(message => s"Error while trying to disable AutoPay: $message")
+      _ <- EitherT(tp.zuoraRestService.cancelSubscription(zuoraSubscription.name)).leftMap(message => s"Error while cancelling subscription: $message")
+      updateReasonResult <- EitherT(tp.zuoraRestService.updateCancellationReason(zuoraSubscription.name, cancelForm.bindFromRequest().value.get)).leftMap(message => s"Error while updating cancellation reason: $message")
+    } yield updateReasonResult).run.map {
+      case -\/(message) =>
+        logger.warn(s"Failed to cancel subscription for user $maybeUserId, due to $message")
+        NotFound()//notFound
```

Comment:
Really there are two cases here, which I'd expect a client of this API to handle differently:

1) We tried to look up the contribution and couldn't find it (404 seems appropriate for this)
2) We found the contribution and tried to cancel it, but something went wrong (404 is not appropriate here)